### ISSUE TITLE
Very slow bootstrap when bundle generated by System builder

### ISSUE THREAD

**Issue by laurentgoudet** [angular/angular#5196]
This [plunker](http://plnkr.co/edit/Q3tmBtiBVjQmcEHqcFoF?p=preview) is a fork of the ["original" Angular plunker](http://plnkr.co/edit/L1zthYrLbD7gkyhPeAew?p=preview), only the Angular version has been updated to the latest (*2.0.0-alpha.45*, unminifed).

This [other fork](http://plnkr.co/edit/t60wzRrPipMTReaXnPP7?p=preview) is based on the first one, expect that the angular2 bundle has been generated by the [System builder](https://github.com/systemjs/builder) through `jspm bundle angular2/angular2 + reflect-metadata` (see below).

For some reason, the later is noticeably slower to bootstrap than the former: "loading..." is shown for several seconds while that the same Angular app bootstrap almost instantaneously in the first plunker. Both bundles have roughly the same size (~1MB).

I see the same issue in my local test app: swapping the System builder bundle (constructed through tracing angular2 ES6 imports) for the distributed bundle (`jspm_packages/npm/angular2@2.0.0-alpha.45/bundles/angular2.dev.js`) saves several seconds of load time.

cc @guybedford, I created the bundle using jspm *0.16.14* is that helps.

```
laurent@MacBook-Pro app [master] $ jspm bundle angular2/angular2 + reflect-metadata
     Building the bundle tree for angular2/angular2 + reflect-metadata...

       github:jspm/nodelibs-process@0.1.2
       github:jspm/nodelibs-process@0.1.2/index
       npm:@reactivex/rxjs@5.0.0-alpha.4/dist/cjs/Observable
       npm:@reactivex/rxjs@5.0.0-alpha.4/dist/cjs/Subject
       npm:@reactivex/rxjs@5.0.0-alpha.4/dist/cjs/Subscriber
       npm:@reactivex/rxjs@5.0.0-alpha.4/dist/cjs/Subscription
       npm:@reactivex/rxjs@5.0.0-alpha.4/dist/cjs/subjects/SubjectSubscription
       npm:@reactivex/rxjs@5.0.0-alpha.4/dist/cjs/util/Symbol_observable
       npm:@reactivex/rxjs@5.0.0-alpha.4/dist/cjs/util/noop
       npm:@reactivex/rxjs@5.0.0-alpha.4/dist/cjs/util/root
       npm:@reactivex/rxjs@5.0.0-alpha.4/dist/cjs/util/throwError
       npm:@reactivex/rxjs@5.0.0-alpha.4/dist/cjs/util/tryOrOnError
       npm:angular2@2.0.0-alpha.45/angular2
       npm:angular2@2.0.0-alpha.45/bootstrap
       npm:angular2@2.0.0-alpha.45/core
       npm:angular2@2.0.0-alpha.45/lifecycle_hooks
       npm:angular2@2.0.0-alpha.45/profile
       npm:angular2@2.0.0-alpha.45/render
       npm:angular2@2.0.0-alpha.45/src/animate/animation
       npm:angular2@2.0.0-alpha.45/src/animate/animation_builder
       npm:angular2@2.0.0-alpha.45/src/animate/browser_details
       npm:angular2@2.0.0-alpha.45/src/animate/css_animation_builder
       npm:angular2@2.0.0-alpha.45/src/animate/css_animation_options
       npm:angular2@2.0.0-alpha.45/src/core/application
       npm:angular2@2.0.0-alpha.45/src/core/application_common
       npm:angular2@2.0.0-alpha.45/src/core/application_ref
       npm:angular2@2.0.0-alpha.45/src/core/application_tokens
       npm:angular2@2.0.0-alpha.45/src/core/bootstrap
       npm:angular2@2.0.0-alpha.45/src/core/change_detection
       npm:angular2@2.0.0-alpha.45/src/core/change_detection/abstract_change_detector
       npm:angular2@2.0.0-alpha.45/src/core/change_detection/binding_record
       npm:angular2@2.0.0-alpha.45/src/core/change_detection/change_detection
       npm:angular2@2.0.0-alpha.45/src/core/change_detection/change_detection_jit_generator
       npm:angular2@2.0.0-alpha.45/src/core/change_detection/change_detection_util
       npm:angular2@2.0.0-alpha.45/src/core/change_detection/change_detector_ref
       npm:angular2@2.0.0-alpha.45/src/core/change_detection/coalesce
       npm:angular2@2.0.0-alpha.45/src/core/change_detection/codegen_facade
       npm:angular2@2.0.0-alpha.45/src/core/change_detection/codegen_logic_util
       npm:angular2@2.0.0-alpha.45/src/core/change_detection/codegen_name_util
       npm:angular2@2.0.0-alpha.45/src/core/change_detection/constants
       npm:angular2@2.0.0-alpha.45/src/core/change_detection/differs/default_iterable_differ
       npm:angular2@2.0.0-alpha.45/src/core/change_detection/differs/default_keyvalue_differ
       npm:angular2@2.0.0-alpha.45/src/core/change_detection/differs/iterable_differs
       npm:angular2@2.0.0-alpha.45/src/core/change_detection/differs/keyvalue_differs
       npm:angular2@2.0.0-alpha.45/src/core/change_detection/directive_record
       npm:angular2@2.0.0-alpha.45/src/core/change_detection/dynamic_change_detector
       npm:angular2@2.0.0-alpha.45/src/core/change_detection/event_binding
       npm:angular2@2.0.0-alpha.45/src/core/change_detection/exceptions
       npm:angular2@2.0.0-alpha.45/src/core/change_detection/interfaces
       npm:angular2@2.0.0-alpha.45/src/core/change_detection/jit_proto_change_detector
       npm:angular2@2.0.0-alpha.45/src/core/change_detection/observable_facade
       npm:angular2@2.0.0-alpha.45/src/core/change_detection/parser/ast
       npm:angular2@2.0.0-alpha.45/src/core/change_detection/parser/lexer
       npm:angular2@2.0.0-alpha.45/src/core/change_detection/parser/locals
       npm:angular2@2.0.0-alpha.45/src/core/change_detection/parser/parser
       npm:angular2@2.0.0-alpha.45/src/core/change_detection/pipe_lifecycle_reflector
       npm:angular2@2.0.0-alpha.45/src/core/change_detection/pipes
       npm:angular2@2.0.0-alpha.45/src/core/change_detection/proto_change_detector
       npm:angular2@2.0.0-alpha.45/src/core/change_detection/proto_record
       npm:angular2@2.0.0-alpha.45/src/core/compiler/anchor_based_app_root_url
       npm:angular2@2.0.0-alpha.45/src/core/compiler/app_root_url
       npm:angular2@2.0.0-alpha.45/src/core/compiler/change_definition_factory
       npm:angular2@2.0.0-alpha.45/src/core/compiler/change_detector_compiler
       npm:angular2@2.0.0-alpha.45/src/core/compiler/command_compiler
       npm:angular2@2.0.0-alpha.45/src/core/compiler/compiler
       npm:angular2@2.0.0-alpha.45/src/core/compiler/directive_metadata
       npm:angular2@2.0.0-alpha.45/src/core/compiler/html_ast
       npm:angular2@2.0.0-alpha.45/src/core/compiler/html_parser
       npm:angular2@2.0.0-alpha.45/src/core/compiler/runtime_compiler
       npm:angular2@2.0.0-alpha.45/src/core/compiler/runtime_metadata
       npm:angular2@2.0.0-alpha.45/src/core/compiler/schema/dom_element_schema_registry
       npm:angular2@2.0.0-alpha.45/src/core/compiler/schema/element_schema_registry
       npm:angular2@2.0.0-alpha.45/src/core/compiler/selector
       npm:angular2@2.0.0-alpha.45/src/core/compiler/shadow_css
       npm:angular2@2.0.0-alpha.45/src/core/compiler/source_module
       npm:angular2@2.0.0-alpha.45/src/core/compiler/style_compiler
       npm:angular2@2.0.0-alpha.45/src/core/compiler/style_url_resolver
       npm:angular2@2.0.0-alpha.45/src/core/compiler/template_ast
       npm:angular2@2.0.0-alpha.45/src/core/compiler/template_compiler
       npm:angular2@2.0.0-alpha.45/src/core/compiler/template_normalizer
       npm:angular2@2.0.0-alpha.45/src/core/compiler/template_parser
       npm:angular2@2.0.0-alpha.45/src/core/compiler/template_preparser
       npm:angular2@2.0.0-alpha.45/src/core/compiler/url_resolver
       npm:angular2@2.0.0-alpha.45/src/core/compiler/util
       npm:angular2@2.0.0-alpha.45/src/core/compiler/xhr
       npm:angular2@2.0.0-alpha.45/src/core/compiler/xhr_impl
       npm:angular2@2.0.0-alpha.45/src/core/debug
       npm:angular2@2.0.0-alpha.45/src/core/debug/debug_element
       npm:angular2@2.0.0-alpha.45/src/core/debug/debug_element_view_listener
       npm:angular2@2.0.0-alpha.45/src/core/di
       npm:angular2@2.0.0-alpha.45/src/core/di/decorators
       npm:angular2@2.0.0-alpha.45/src/core/di/exceptions
       npm:angular2@2.0.0-alpha.45/src/core/di/forward_ref
       npm:angular2@2.0.0-alpha.45/src/core/di/injector
       npm:angular2@2.0.0-alpha.45/src/core/di/key
       npm:angular2@2.0.0-alpha.45/src/core/di/metadata
       npm:angular2@2.0.0-alpha.45/src/core/di/opaque_token
       npm:angular2@2.0.0-alpha.45/src/core/di/provider
       npm:angular2@2.0.0-alpha.45/src/core/di/type_literal
       npm:angular2@2.0.0-alpha.45/src/core/directives
       npm:angular2@2.0.0-alpha.45/src/core/directives/ng_class
       npm:angular2@2.0.0-alpha.45/src/core/directives/ng_for
       npm:angular2@2.0.0-alpha.45/src/core/directives/ng_if
       npm:angular2@2.0.0-alpha.45/src/core/directives/ng_style
       npm:angular2@2.0.0-alpha.45/src/core/directives/ng_switch
       npm:angular2@2.0.0-alpha.45/src/core/directives/observable_list_diff
       npm:angular2@2.0.0-alpha.45/src/core/dom/browser_adapter
       npm:angular2@2.0.0-alpha.45/src/core/dom/dom_adapter
       npm:angular2@2.0.0-alpha.45/src/core/dom/generic_browser_adapter
       npm:angular2@2.0.0-alpha.45/src/core/facade
       npm:angular2@2.0.0-alpha.45/src/core/facade/async
       npm:angular2@2.0.0-alpha.45/src/core/facade/collection
       npm:angular2@2.0.0-alpha.45/src/core/facade/exception_handler
       npm:angular2@2.0.0-alpha.45/src/core/facade/exceptions
       npm:angular2@2.0.0-alpha.45/src/core/facade/intl
       npm:angular2@2.0.0-alpha.45/src/core/facade/lang
       npm:angular2@2.0.0-alpha.45/src/core/facade/math
       npm:angular2@2.0.0-alpha.45/src/core/facade/promise
       npm:angular2@2.0.0-alpha.45/src/core/forms
       npm:angular2@2.0.0-alpha.45/src/core/forms/directives
       npm:angular2@2.0.0-alpha.45/src/core/forms/directives/abstract_control_directive
       npm:angular2@2.0.0-alpha.45/src/core/forms/directives/checkbox_value_accessor
       npm:angular2@2.0.0-alpha.45/src/core/forms/directives/control_container
       npm:angular2@2.0.0-alpha.45/src/core/forms/directives/control_value_accessor
       npm:angular2@2.0.0-alpha.45/src/core/forms/directives/default_value_accessor
       npm:angular2@2.0.0-alpha.45/src/core/forms/directives/ng_control
       npm:angular2@2.0.0-alpha.45/src/core/forms/directives/ng_control_group
       npm:angular2@2.0.0-alpha.45/src/core/forms/directives/ng_control_name
       npm:angular2@2.0.0-alpha.45/src/core/forms/directives/ng_control_status
       npm:angular2@2.0.0-alpha.45/src/core/forms/directives/ng_form
       npm:angular2@2.0.0-alpha.45/src/core/forms/directives/ng_form_control
       npm:angular2@2.0.0-alpha.45/src/core/forms/directives/ng_form_model
       npm:angular2@2.0.0-alpha.45/src/core/forms/directives/ng_model
       npm:angular2@2.0.0-alpha.45/src/core/forms/directives/normalize_validator
       npm:angular2@2.0.0-alpha.45/src/core/forms/directives/number_value_accessor
       npm:angular2@2.0.0-alpha.45/src/core/forms/directives/select_control_value_accessor
       npm:angular2@2.0.0-alpha.45/src/core/forms/directives/shared
       npm:angular2@2.0.0-alpha.45/src/core/forms/directives/validators
       npm:angular2@2.0.0-alpha.45/src/core/forms/form_builder
       npm:angular2@2.0.0-alpha.45/src/core/forms/model
       npm:angular2@2.0.0-alpha.45/src/core/forms/validators
       npm:angular2@2.0.0-alpha.45/src/core/life_cycle/life_cycle
       npm:angular2@2.0.0-alpha.45/src/core/lifecycle
       npm:angular2@2.0.0-alpha.45/src/core/linker
       npm:angular2@2.0.0-alpha.45/src/core/linker/compiler
       npm:angular2@2.0.0-alpha.45/src/core/linker/directive_lifecycle_reflector
       npm:angular2@2.0.0-alpha.45/src/core/linker/directive_resolver
       npm:angular2@2.0.0-alpha.45/src/core/linker/dynamic_component_loader
       npm:angular2@2.0.0-alpha.45/src/core/linker/element_binder
       npm:angular2@2.0.0-alpha.45/src/core/linker/element_injector
       npm:angular2@2.0.0-alpha.45/src/core/linker/element_ref
       npm:angular2@2.0.0-alpha.45/src/core/linker/event_config
       npm:angular2@2.0.0-alpha.45/src/core/linker/interfaces
       npm:angular2@2.0.0-alpha.45/src/core/linker/pipe_resolver
       npm:angular2@2.0.0-alpha.45/src/core/linker/proto_view_factory
       npm:angular2@2.0.0-alpha.45/src/core/linker/query_list
       npm:angular2@2.0.0-alpha.45/src/core/linker/template_commands
       npm:angular2@2.0.0-alpha.45/src/core/linker/template_ref
       npm:angular2@2.0.0-alpha.45/src/core/linker/view
       npm:angular2@2.0.0-alpha.45/src/core/linker/view_container_ref
       npm:angular2@2.0.0-alpha.45/src/core/linker/view_listener
       npm:angular2@2.0.0-alpha.45/src/core/linker/view_manager
       npm:angular2@2.0.0-alpha.45/src/core/linker/view_manager_utils
       npm:angular2@2.0.0-alpha.45/src/core/linker/view_pool
       npm:angular2@2.0.0-alpha.45/src/core/linker/view_ref
       npm:angular2@2.0.0-alpha.45/src/core/linker/view_resolver
       npm:angular2@2.0.0-alpha.45/src/core/metadata
       npm:angular2@2.0.0-alpha.45/src/core/metadata/di
       npm:angular2@2.0.0-alpha.45/src/core/metadata/directives
       npm:angular2@2.0.0-alpha.45/src/core/metadata/view
       npm:angular2@2.0.0-alpha.45/src/core/pipes
       npm:angular2@2.0.0-alpha.45/src/core/pipes/async_pipe
       npm:angular2@2.0.0-alpha.45/src/core/pipes/date_pipe
       npm:angular2@2.0.0-alpha.45/src/core/pipes/default_pipes
       npm:angular2@2.0.0-alpha.45/src/core/pipes/invalid_pipe_argument_exception
       npm:angular2@2.0.0-alpha.45/src/core/pipes/json_pipe
       npm:angular2@2.0.0-alpha.45/src/core/pipes/lowercase_pipe
       npm:angular2@2.0.0-alpha.45/src/core/pipes/number_pipe
       npm:angular2@2.0.0-alpha.45/src/core/pipes/pipe_provider
       npm:angular2@2.0.0-alpha.45/src/core/pipes/pipes
       npm:angular2@2.0.0-alpha.45/src/core/pipes/slice_pipe
       npm:angular2@2.0.0-alpha.45/src/core/pipes/uppercase_pipe
       npm:angular2@2.0.0-alpha.45/src/core/platform_bindings
       npm:angular2@2.0.0-alpha.45/src/core/profile/profile
       npm:angular2@2.0.0-alpha.45/src/core/profile/wtf_impl
       npm:angular2@2.0.0-alpha.45/src/core/profile/wtf_init
       npm:angular2@2.0.0-alpha.45/src/core/reflection/reflection
       npm:angular2@2.0.0-alpha.45/src/core/reflection/reflection_capabilities
       npm:angular2@2.0.0-alpha.45/src/core/reflection/reflector
       npm:angular2@2.0.0-alpha.45/src/core/render
       npm:angular2@2.0.0-alpha.45/src/core/render/api
       npm:angular2@2.0.0-alpha.45/src/core/render/dom/dom_renderer
       npm:angular2@2.0.0-alpha.45/src/core/render/dom/dom_tokens
       npm:angular2@2.0.0-alpha.45/src/core/render/dom/events/event_manager
       npm:angular2@2.0.0-alpha.45/src/core/render/dom/events/hammer_common
       npm:angular2@2.0.0-alpha.45/src/core/render/dom/events/hammer_gestures
       npm:angular2@2.0.0-alpha.45/src/core/render/dom/events/key_events
       npm:angular2@2.0.0-alpha.45/src/core/render/dom/shared_styles_host
       npm:angular2@2.0.0-alpha.45/src/core/render/dom/util
       npm:angular2@2.0.0-alpha.45/src/core/render/render
       npm:angular2@2.0.0-alpha.45/src/core/render/view
       npm:angular2@2.0.0-alpha.45/src/core/render/view_factory
       npm:angular2@2.0.0-alpha.45/src/core/services
       npm:angular2@2.0.0-alpha.45/src/core/services/title
       npm:angular2@2.0.0-alpha.45/src/core/testability/browser_testability
       npm:angular2@2.0.0-alpha.45/src/core/testability/testability
       npm:angular2@2.0.0-alpha.45/src/core/util
       npm:angular2@2.0.0-alpha.45/src/core/util/decorators
       npm:angular2@2.0.0-alpha.45/src/core/zone
       npm:angular2@2.0.0-alpha.45/src/core/zone/ng_zone
       npm:angular2@2.0.0-alpha.45/src/transform/template_compiler/change_detector_codegen
       npm:process@0.11.2
       npm:process@0.11.2/browser
       npm:reflect-metadata@0.1.2
       npm:reflect-metadata@0.1.2/Reflect

ok   Built into build.js with source maps, unminified.
```

**Comment by laurentgoudet** [comment]
Ok I'm going to sound crazy but after further testing - I was trying to look at the timeline - I've realized that it only happens when the Chrome DevTools are opened 0_o. I've grabbed a screencast of the issue https://youtu.be/e1cTPIDCCZE, which I can reproduce using the previous plunkers in both Chrome stable v46 and Canary v48. Even worse, trying the capture a timeline "fixes" the issue: no more visible lag with the System builder's bundle.

**Comment by pkozlowski-opensource** [comment]
@laurentgoudet currently there are way too many tools / workflows to create bundles from ES6-based modules. Unfortunately we can't investigate all the possible combinations and quirks of each approach. At least for now if you go outside of the recommended territory you meed to be prepared to debug / troubleshoot things on your own. 

Having said this we want to make sure that all major workflows are supported, so if you spot anything that we do on ng2 preventing your favorite workflow - please do let us know. But once again, you need to be ready to investigate on your end.

**Comment by laurentgoudet** [comment]
Thanks @pkozlowski-opensource. Yeah I've seen your Angular Connect talk on the subject - great one by the way. I did some more tests locally: I had Async debugging enabled in the DevTools, which makes the issue way worse. Without it, the `System.register` bundle is still consistently 30% to 50% slower than the CommonJS one (1s vs. 1.5s fully loaded), only when the DevTools are opened (JS source maps are disabled).

My first guess would be that it's due to the `System.register` circular reference "correctness" overhead, however I can't explain why I'm only seeing that when the DevTools are enabled. Maybe the `System.register` format lead to deeper call stacks which makes the JS instrumentation more costly.

Anyway, I'm closing this issue since it doesn't seem to be Angular specific, and I've created a new one of the SystemJS tracker: https://github.com/systemjs/systemjs/issues/908
